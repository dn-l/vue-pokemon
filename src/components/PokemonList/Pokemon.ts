interface Attack {
  damage: number;
}

interface Attacks {
  fast: Attack[];
  special: Attack[];
}

interface Pokemon {
  id: string;
  name: string;
  number: string;
  evolutions: Pokemon[];
  isLoading: boolean;
  attacks: Attacks;
}

export default Pokemon;
