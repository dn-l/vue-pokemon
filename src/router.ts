import Vue from 'vue';
import Router from 'vue-router';
import Pokemons from './views/Pokemons.vue';

export const routes = [
  {
    path: '/',
    name: 'pokemons',
    component: Pokemons,
  },
  {
    path: '/pokemon/:id',
    name: 'pokemon',
    component: () => import(/* webpackChunkName: "pokemon" */ './views/Pokemon.vue'),
  },
  {
    path: '*',
    redirect: '/',
  },
];

Vue.use(Router);

export default new Router({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});
