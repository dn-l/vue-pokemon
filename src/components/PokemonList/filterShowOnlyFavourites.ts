import Pokemon from './Pokemon';

export default (
  pokemons: Pokemon[], favourites: string[],
) => pokemons.filter(({ id }) => favourites.includes(id));
