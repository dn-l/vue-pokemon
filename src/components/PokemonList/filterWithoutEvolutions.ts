import Pokemon from './Pokemon';

export default (pokemons: Pokemon[]) => {
  const pokemonIsNotAnEvolution = (pokemon: Pokemon) => pokemons
    .every(({ evolutions }) => !evolutions
      || evolutions.every(({ id }) => id !== pokemon.id));
  return pokemons.filter(pokemonIsNotAnEvolution);
};
