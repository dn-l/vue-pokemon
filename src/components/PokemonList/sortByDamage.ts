import Pokemon from './Pokemon';

const getPokemonDamage = (
  { attacks: { fast, special } }: Pokemon,
) => Math.max(...[
  ...fast.map(({ damage }) => damage),
  ...special.map(({ damage }) => damage),
]);

export default (pokemons: Pokemon[]) => pokemons
  .sort((a: Pokemon, b: Pokemon) => getPokemonDamage(a) - getPokemonDamage(b));
