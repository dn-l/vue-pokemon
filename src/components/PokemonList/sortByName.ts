import Pokemon from './Pokemon';

export default (pokemons: Pokemon[]) => pokemons
  .sort((a: Pokemon, b: Pokemon) => a.name.localeCompare(b.name));
