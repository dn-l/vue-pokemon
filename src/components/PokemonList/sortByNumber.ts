import Pokemon from './Pokemon';

export default (pokemons: Pokemon[]) => pokemons
  .sort((a: Pokemon, b: Pokemon) => Number(a.number) - Number(b.number));
