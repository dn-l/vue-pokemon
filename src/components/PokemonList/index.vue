<template>
  <v-row justify="space-around" align="center">
    <PokemonCard :pokemon="pokemon" :key="pokemon.id" v-for="pokemon in filteredPokemons" />
  </v-row>
</template>

<script lang="ts">
import Vue from 'vue';
import PokemonCard from '../PokemonCard.vue';
import Pokemon from './Pokemon';
import filterShowOnlyFavourites from './filterShowOnlyFavourites';
import filterWithoutEvolutions from './filterWithoutEvolutions';
import sortByName from './sortByName';
import sortByNumber from './sortByNumber';
import sortByDamage from './sortByDamage';
import State, { SortBy } from '../../State';

export default Vue.extend({
  props: ['pokemons'],

  components: {
    PokemonCard,
  },

  computed: {
    filteredPokemons() {
      let pokemons = this.$props.pokemons as Pokemon[];
      const {
        favourites,
        filterParams: {
          showOnlyFavourite,
          showEvolutions,
        },
        sortBy,
      } = this.$store.state as State;

      if (!showEvolutions) {
        pokemons = filterWithoutEvolutions(pokemons);
      }

      if (showOnlyFavourite) {
        pokemons = filterShowOnlyFavourites(pokemons, favourites);
      }

      switch (sortBy) {
        case SortBy.Number:
          return sortByNumber(pokemons);

        case SortBy.Damage:
          return sortByDamage(pokemons);

        // SortBy.Name
        default:
          return sortByName(pokemons);
      }
    },
  },
});
</script>
