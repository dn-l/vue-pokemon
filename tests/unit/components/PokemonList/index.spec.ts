import Vue from 'vue';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import { routes } from '@/router';
import PokemonList from '@/components/PokemonList/index.vue';

const pokemons = [
  { id: '1', name: 'Pikachu' },
  { id: '2', name: 'Bulbasaur' },
];

const localVue = createLocalVue();
Vue.use(Vuetify);
localVue.use(Vuex);
localVue.use(VueRouter);

const router = new VueRouter({ routes });

const store = new Vuex.Store({
  state: { filterParams: { } },
});

describe('PokemonList.vue', () => {
  const wrapper = shallowMount(PokemonList, {
    propsData: { pokemons },
    store,
    localVue,
    router,
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
