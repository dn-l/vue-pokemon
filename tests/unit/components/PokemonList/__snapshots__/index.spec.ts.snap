// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonList.vue matches snapshot 1`] = `
<v-row-stub tag="div" align="center" justify="space-around">
  <pokemoncard-stub pokemon="[object Object]"></pokemoncard-stub>
  <pokemoncard-stub pokemon="[object Object]"></pokemoncard-stub>
</v-row-stub>
`;
